functions/ts/src/sops.ts

```diff
@@ -0,0 +1,223 @@
+  }
+
+  // Put all documents into the local place and
+  // cleanup the configs storage. This is to keep
+  // the order of documents
+  const allDocs = configs.getAll();
+  configs.deleteAll();
+
+  for (const object of allDocs) {
+    if (isSopsKubernetesObject(object)) {
+      // this function
+      await decryptAndInsertSopsKubernetesObject(args, configs, object);
+    } else {
+      // add it back
+      configs.insert(object);
+    }
+  }
+}
+
+function detachAnnotations(object: SopsKubernetesObject): Map<string, string> {
+  const detached = new Map<string, string>();
+
+  for (const annotation of detachedAnnotations) {
+    const value = getAnnotation(object, annotation);
+    if (value !== undefined) {
+      detached.set(annotation, value);
+      removeAnnotation(object, annotation);
+    }
+  }
+  return detached;
+}
+
+function attachDetachedAnnotations(
+  object: SopsKubernetesObject,
+  detached: Map<string, string>
+) {
+  detached.forEach((value: string, key: string) => {
+    addAnnotation(object, key, value);
+  });
+}
+
+async function decryptAndInsertSopsKubernetesObject(
+  args: string[],
+  configs: Configs,
+  object: SopsKubernetesObject
+) {
+  let error;
+  args.push(...['-d', TEMP_PATH]);
+
+  // write encrypted file to the temp file
+  const detached = detachAnnotations(object);
+  const stringifiedObject = safeDump(object, YAML_STYLE);
+  await writeFile(TEMP_PATH, stringifiedObject);
+  attachDetachedAnnotations(object, detached);
+
+  try {
+    // run sops
+    const child = spawnSync('sops', args);
+    error = child.stderr;
+    // read the decrypted yaml from stdout and parse
+    const decryptedObject = safeLoad(child.stdout);
+    if (object && isKubernetesObject(decryptedObject)) {
+      attachDetachedAnnotations(decryptedObject, detached);
+      configs.insert(decryptedObject);
+      return;
+    }
+  } catch (err) {
+    configs.addResults(
+      generalResult(
+        `Exception for apiVersion: ${object.apiVersion}, kind: ${object.kind}, name: ${object.metadata.name}: ${err}`,
+        'error'
+      )
+    );
+  }
+  if (error && error.length > 0) {
+    configs.addResults(
+      generalResult(
+        `Sops command results in error for\n${stringifiedObject}\n ${error.toString()}`,
+        'error'
+      )
+    );
+  }
+  // if we're here - there was an error
+  // putting the original object back
+  configs.insert(object);
+}
+
+function readSopsArguments(configs: Configs) {
+  const args: string[] = [];
+  const configMap = configs.getFunctionConfigMap();
+  if (!configMap) {
+    return args;
+  }
+  configMap.forEach((value: string, key: string) => {
+    if (key === OVERRIDE_PREEXEC_CMD) {
+      preExecCmd = value;
+    } else if (key === OVERRIDE_DETACHED_ANNOTATIONS) {
+      detachedAnnotations = value.replace(/\s/g, '').split(',');
+    } else if (key === VERBOSE || key === IGNORE_MAC) {
+      args.push('--' + key);
+    } else {
+      args.push('--' + key);
```

Comment:
Nice way to handle arbitrary flags! We've had issues with this in the past (https://github.com/GoogleContainerTools/kpt/issues/824)